qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -168,5 +208,233 @@ class DefaultSynthesisLinearFunction(HighLevelSynthesisPlugin):
 
     def run(self, high_level_object, **options):
         """Run synthesis for the given LinearFunction."""
-        decomposition = synth_cnot_count_full_pmh(high_level_object.linear)
+        # For now, use PMH algorithm by default
+        decomposition = PMHSynthesisLinearFunction().run(high_level_object, **options)
         return decomposition
+
+
+class KMSSynthesisLinearFunction(HighLevelSynthesisPlugin):
+    """Linear function synthesis plugin based on the Kutin-Moulton-Smithline method.
+
+    The plugin also supports the following plugin-specific options that can be passed
+    in ``__init__`` and ``run`` methods:
+
+    * coupling_map: CouplingMap of the target backend or ``None``. If this coupling map
+      is not ``None``, then the synthesized quantum circuit should adhere to the imposed
+      connectivity constraints. Note that the synthesis algorithm is allowed to return
+      ``None`` when it is unable to synthesize the desired circuit.
+    * qubits: the sequence of qubits over which the ``high_level_object`` is defined.
+      This is only used in the case that the coupling map is not ``None``.
+    * opt_count: an option to prioritize count over depth when multiple synthesized
+      circuits are available.
+    * all_mats: an option to synthesize quantum circuits for the matrix, its transpose,
+      its inverse, and its inverse transpose, choosing the best implementation.
+    * max_paths: when the coupling map is not ``None``, the internally used synthesis
+      algorithm needs to pick a hamiltonian path through ``qubits``. By default, only one
+      hamiltonian pass is considered. This option allows to increase the maximum number
+      of paths that should be considered, choosing the best implementation.
+    * orig_circuit: an option to also consider the original implementation of the linear
+      function when available, choosing the best implementation. Note that the original
+      implementation might not adhere to the coupling map.
+
+    The output of the ``run`` method should either be ``None`` if the circuit could
+    not be synthesized, a ``QuantumCircuit``, or a pair consisting of a quantum circuit
+    and an ordered list of qubits.
+    """
+
+    def __init__(self, **options):
+        self._options = options
```

Comment:
I'm not a huge fan of this for things that are not directly user specified via the `HLSConfig`. I'm thinking `coupling_map` is different here because it's a target constraint and just bundling it via manually specified options by the user (in the `HLSConfig`) isn't the best interface for plugins. I feel like we really need something like what unitary synthesis is doing that defines the interface options vs the user specified configuration. Right now if I wanted to write a hardware aware synthesis plugin I wouldn't know what is a standard feature of the interface and might end up diverging from the standard interface.